wip-asdpc-rng: submit draft (#70)
witnet/WIPs: wip-0019.md

<pre>
  WIP: WIP-0019
  Layer: Consensus (hard fork)
  Title: Random Number Generation (RNG) functionality
  Authors: Adán SDPC <adan@witnet.foundation>
  Discussions-To: https://github.com/witnet/witnet-rust/discussions/2058
  Status: Draft
  Type: Standards Track
  Created: 2021-09-06
  License: BSD-2-Clause
</pre>

## Abstract

> A new type of data source is proposed such that leverages the crowd-attestation mechanism of the Witnet protocol

Reviewer: The **such that leverages** sounds weird to me. Maybe something like this?

```suggestion
A new type of data source is proposed to leverage the crowd-attestation mechanism of the Witnet protocol to enable Random Number Generation (RNG) functionality inside data requests.
```

Author: Too British for a WIP, my bad! :rofl:

to enable Random Number Generation (RNG) functionality inside data requests.

## Motivation and rationale

Just in the same way as price feeds are the fundamental oracle-based building block for many DeFi protocols RNGs are
quickly becoming a keystone building block for non-fungible tokens (NFTs).  

Many use cases are arising in which the smart contracts generating the NFTs are in need of reliable and tamper-proof
random inputs for their generative algorithms ensuring the "uniqueness" of each piece. That is, they need to randomly
assign a distinguishable set of scarce attributes or traits to each newly minted NFT unit. 

The case for oracle-powered RNGs expands beyond NFTs, as many other constructs also benefit from random inputs, e.g.
in the area of gaming, lotteries, gambling, random committee selection, etc.

Due to the robustness and flexibility of the crowd-attestation mechanism of the Witnet protocol, there is a huge
opportunity to add RNG capabilities to it with very little and self-contained changes.

> In particular, the mechanism hereby introduced consists on a new type of data source in which no external HTTP(S)

Reviewer: ```suggestion
In particular, the mechanism hereby introduced consists of a new type of data source in which no external HTTP(S)
```

servers are queried, and nodes are instead required to secretly commit secret sequences of bytes. The rest of the
protocol follows in the traditional way, with the only addition of a new reducer function that concatenates the
revealed values in a deterministic order and hashes the sequence together so as to produce a single, uniform and
pseudo-random output that will be used as the result of the request.

Additionally, a new deterministic ordering for reveals inside tally transactions is introduced so as to rule out any
malleability attempt from tally producers. 

Due to the overlap of Witnet's inherent commitment scheme with this last aggregation step, there is no need for the
committed bytes to be verifiably random. Under the assumptions of stock hashing functions, even if the committers
collude around crafting their commitments in such a way that the output is favorable to them, as long as 1 single
committer remains independent and commits a value that is secret to the rest, the final result will hold the desired
randomness properties.

> 

Reviewer: The other very unlikely attack would be a DoS attack in which a `REVEAL` transaction is "denied" during the  `reveal_extra_rounds` period. For that to happen, an attacker would have to be able to mine 4 blocks consecutively.

Author: I included a paragraph stating this more explicitely

## Specification

### Introduction of data request type `1`

**(1)** A new type of data source is introduced with name `Rng` and identifier `1`.

**(2)** Type `1` sources have the same `script` attribute as type `0` sources, but no `url` attribute.

**(3)** The script field in type `1` sources use the same RADON operators and encoding as type `0`.

### Retrieval of type `1` sources

**(4)** When retrieving a type `1` source, a node MUST generate on the spot a random and secret sequence of 32
bytes.

> **(5)** This random and secret sequence of 32 bytes MUST be used as the result of the retrieval. 

Reviewer 2: What should happen when there is a retrieval script? Should that script be executed or ignored? This sentence seems to imply that it should be ignored?

Author: In the absence of any specific mention, everything follows the original logic and the script should be executed.

Author: Do we want to forbid scripts on RNG sources? I guess there's not much use to them, but who knows.

Reviewer 2: If we forbid scripts we can save 1 byte because `[]` is not a valid script, we currently must use `[128]`. Other than that, the only use case I see for allowing scripts would be to try to write a malicious request that looks like a random request but the script makes it deterministic.

Reviewer 3: I can see a use for scripts where you'd process the random number on the Witnet blockchain already thereby limiting the number of calculations that need to happen in the smart contract? Currently I don't think enough operators are supported to do useful processing, but that may change in the future.

Author: I'm unsure about the point of doing any kind of operation on the "local" randomness, that is, before aggregating the reveals form multiple witnesses.

A different question would be if an script could be provided such that it is applied on the result upon tally. I do believe that would be more in line which was drcpu has in mind (e.g. going from uniform to normal, reducing the randomness to a specific range, etc.)

Reviewer 3: Yes, I'm talking about a tally script, I missed the context of this being at the aggregation step (though I'm not sure if it would harm being able to do this in the aggregation stage too).

Author: > If we forbid scripts we can save 1 byte because `[]` is not a valid script, we currently must use `[128]`. Other than that, the only use case I see for allowing scripts would be to try to write a malicious request that looks like a random request but the script makes it deterministic.

Nice point. Btw in a future WIP we could introduce implicit typing of scripts and other structures (remove the first byte because it's always `128`)

Reviewer 2: > I'm unsure about the point of doing any kind of operation on the "local" randomness, that is, before aggregating the reveals form multiple witnesses.

If the computation happens before the commit-reveal steps, the result is impossible to prove and therefore insecure. The early versions of witnet used to allow arbitrary scripts in the tally stage, but that was removed for some reason I don't remember, most likely for simplicity.

Author: I don't remember either why we gave up on tally scripts, but I must recognize at this point I don't see any issue with them other than additional complexity.

**(6)** Data requests containing type `1` sources MUST be revealed normally.

### Ordering of reveals in tally transactions

**(7)** When producing and validating tally transactions, reveals must be listed and considered in increasing order
of the big-endian value of the `SHA256` hash of the concatenation of each revealer's 20-bytes Public Key Hash followed
by the 32-bytes ID of the data request.

```rust
// Rust-alike pseudocode
reveals.sort(|reveal| {
  let mut data = Vec::new();
  data.extend(reveal.body.pkh.bytes);
  data.extend(data_request_id.bytes);

  sha256(data)
});
```

### Addition of a `ConcatenateAndHash` reducer

**(8)** A new reducer function with code `0x0B` and name `ConcatenateAndHash` is introduced.

**(9)** The `ConcatenateAndHash` reducer function is a general purpose reducer function: it can be used as a reducer
in the RADON scripts of any type of data sources, as well as in the reducing step of the aggregation and tally stages.

**(10)** The `ConcatenateAndHash` reducer function only accepts a non-empty homogeneous `Array` of byte sequences as its
input. That is, an `Array` containing one or more items of the `Bytes` type.

> **(11)** May the `ConcatenateAndHash` reducer function be applied on an input not satisfying (9), output MUST be an

Reviewer 2: I don't understand what does `(9)` refer to. Perhaps it's a typo?

```suggestion
**(11)** May the `ConcatenateAndHash` reducer function be applied on an input not satisfying (10), output MUST be an
```

Author: True, I'll fix it. Thanks!

Author: Resolved

- Non-implementer: a client that fails to or refuses to implement the protocol improvements proposed in this document.

#### Consensus cheat sheet

Upon entry into force of the proposed improvements:

- Blocks and transactions that were considered valid by former versions of the protocol MUST continue to be considered valid.
- Blocks and transactions produced by non-implementers MUST be validated by implementers using the same rules as with those coming from implementers, that is, the new rules.
- Implementers MUST apply the proposed logic when evaluating transactions or computing their own data request eligibility.

As a result:

- Non-implementers MAY NOT get their blocks and transactions accepted by implementers.
- Implementers MAY NOT get their valid blocks accepted by non-implementers.
- Non-implementers MAY consolidate different blocks and superblocks than implementers.

Due to this last point, this MUST be considered a consensus-critical protocol improvement. An adoption plan MUST be proposed, setting an activation date that gives miners enough time in advance to upgrade their existing clients.

### Libraries, clients, and interfaces

> The protocol improvements proposed herein will affect any library or client implementing the any functionality related to the RAD engine.

Reviewer: ```suggestion
The protocol improvements proposed herein will affect any library or client implementing any functionality related to the RAD engine.
```
